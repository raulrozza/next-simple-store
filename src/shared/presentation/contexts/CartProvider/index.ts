export { default as useAddToCart } from './hooks/useAddToCart';
