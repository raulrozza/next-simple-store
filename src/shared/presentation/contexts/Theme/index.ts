export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
