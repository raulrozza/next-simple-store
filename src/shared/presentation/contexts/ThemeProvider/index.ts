export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
