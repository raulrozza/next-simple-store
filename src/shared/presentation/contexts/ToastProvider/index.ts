export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
