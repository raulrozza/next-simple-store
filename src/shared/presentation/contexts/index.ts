export * from './CartProvider';
export * from './ThemeProvider';
