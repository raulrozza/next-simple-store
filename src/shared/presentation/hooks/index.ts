import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
