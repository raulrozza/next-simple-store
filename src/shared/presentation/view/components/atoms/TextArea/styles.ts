import styled, { css } from 'styled-components';

interface TextAreaProps {
    hasErrors?: boolean;
}

export const StyledTextArea = styled.textarea<TextAreaProps>`
    ${({ theme, hasErrors }) => css`
        background-color: ${theme.palette.gray['0']};
        border: 2px solid ${theme.palette.gray['100']};
        outline: none;
        border-radius: ${theme.layout.borderRadius.sm};

        width: 100%;
        height: ${theme.layout.size(6)};
        padding: ${theme.layout.spacing(2)};
        transition: outline 0.2s, border 0.2s;

        resize: none;

        &:focus {
            border-color: ${theme.palette.primary['400']};
        }

        ${hasErrors &&
        css`
            border-color: ${theme.palette.error['400']};

            &:focus {
                border-color: ${theme.palette.error['200']};
            }
        `}
    `}
`;

export const ErrorContainer = styled.span`
    ${({ theme }) => css`
        color: ${theme.palette.error['400']};
        font-size: ${theme.typography.sizes.sm};
        padding: ${theme.layout.spacing(1, 1, 0, 1)};
    `}
`;
