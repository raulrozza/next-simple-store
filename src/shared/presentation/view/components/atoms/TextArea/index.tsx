import { FC } from 'react';

import { useField } from 'formik';

import { ITextArea } from '@/shared/presentation/entities/Field';

import { ErrorContainer, StyledTextArea } from './styles';

const TextArea: FC<ITextArea> = ({ name, ...props }) => {
    const [field, meta] = useField(name);

    const hasErrors = Boolean(meta.touched && meta.error);

    return (
        <div>
            <StyledTextArea {...props} {...field} hasErrors={hasErrors} />

            {hasErrors && <ErrorContainer>{meta.error}</ErrorContainer>}
        </div>
    );
};

export default TextArea;
