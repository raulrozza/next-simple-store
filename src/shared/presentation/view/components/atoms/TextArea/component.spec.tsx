import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Formik } from 'formik';

import { ThemeProvider } from '@/shared/presentation/contexts';

import TextArea from './index';

describe('TextArea', () => {
    it('should render the textarea correctly', () => {
        const view = render(
            <Formik
                initialValues={{
                    textarea: '',
                }}
                onSubmit={jest.fn()}
            >
                <TextArea name="textarea" placeholder="My textarea" />
            </Formik>,
            {
                wrapper: ThemeProvider,
            },
        );

        expect(screen.getByPlaceholderText('My textarea')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('My textarea')).toHaveValue('');
        expect(view).toMatchSnapshot();
    });

    describe('Value handling', () => {
        it('should init the textarea with the initial field value', () => {
            const view = render(
                <Formik
                    initialValues={{ textarea: 'my value' }}
                    onSubmit={jest.fn()}
                >
                    <TextArea name="textarea" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            expect(screen.getByDisplayValue(/my value/)).toHaveValue(
                'my value',
            );
            expect(view).toMatchSnapshot();
        });

        it('should change the field value when typing the textarea', async () => {
            render(
                <Formik
                    initialValues={{ textarea: 'my value' }}
                    onSubmit={jest.fn()}
                >
                    <TextArea name="textarea" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            const textarea = screen.getByDisplayValue(/my value/);
            await userEvent.type(textarea, 'new value');

            expect(screen.getByDisplayValue(/new value/)).toBeDefined();
        });
    });

    describe('Error handling', () => {
        it('should show the error message if the textarea has errors and has been touched', async () => {
            const user = userEvent.setup();

            render(
                <Formik
                    initialValues={{
                        textarea: '',
                    }}
                    validate={() => ({ textarea: 'My error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <TextArea name="textarea" placeholder="My textarea" />
                </Formik>,
                {
                    wrapper: ThemeProvider,
                },
            );

            expect(screen.queryByText('My error')).not.toBeInTheDocument();

            await user.type(
                screen.getByPlaceholderText('My textarea'),
                'new value',
            );
            act(() => {
                screen.getByPlaceholderText('My textarea').blur(); // Deselect to trigger touch
            });

            expect(await screen.findByText('My error')).toBeInTheDocument();
        });

        it('should not display the error message if the textarea was not touched', async () => {
            render(
                <Formik
                    initialValues={{ textarea: 'my value' }}
                    validate={() => ({ textarea: 'my error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <TextArea name="textarea" placeholder="My textarea" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            await userEvent.type(
                screen.getByPlaceholderText('My textarea'),
                'invalid text',
            );

            expect(screen.queryByText(/my error/)).toBeNull();
        });
    });
});
