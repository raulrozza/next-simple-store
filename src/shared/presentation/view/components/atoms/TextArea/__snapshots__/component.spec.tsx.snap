// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Value handling should init the textarea with the initial field value 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  height: 6rem;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
  resize: none;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <textarea
          class="c0"
          name="textarea"
        >
          my value
        </textarea>
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  height: 6rem;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
  resize: none;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <textarea
        class="c0"
        name="textarea"
      >
        my value
      </textarea>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TextArea should render the textarea correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  height: 6rem;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
  resize: none;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <textarea
          class="c0"
          name="textarea"
          placeholder="My textarea"
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  height: 6rem;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
  resize: none;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <textarea
        class="c0"
        name="textarea"
        placeholder="My textarea"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
